compiler/rustc_arena/src/lib.rs

```diff
@@ -55,15 +56,22 @@ pub struct TypedArena<T> {
 
 struct ArenaChunk<T = u8> {
     /// The raw storage for the arena chunk.
-    storage: Box<[MaybeUninit<T>]>,
+    /// This is a `*const T` and not a `*mut T` because we own the value.
```

Comment:
Maybe this is a stupid question, but why is a const pointer better here than a mut?

Reply:
The only difference between the two pointer types is variance. `*mut T` is invariant over `T` (just like `&mut T`), and `*const T` is covariant over `T` (just like `&T`). This makes it obvious which pointer types should be used as an analog to `&T` and `&mut T`. An owned value presents a more interesting problem. For owned values, covariance (which is more permissive than invariance) is ok, and it's also the variance of `Box<T>` (which was previously used here). So we probably want to keep the covariance for flexibility (though I doubt that it matters here). So I used `*const T` here. Though I just realized that `NonNull<T>` would be better, since it's also covariant and also gets the nullpointer niche,  though I haven't checked whether that would be useful here.

Reply:
To learn more about variance, check out the variance chapter in the Rustonomicon (I recommend reading the new version [here](https://github.com/rust-lang/nomicon/pull/340)).

Reply:
+1 to using `NonNull<T>` here